From a8bc0f90ef49ea0aee90047a17772e4eebff259a Mon Sep 17 00:00:00 2001
From: Sandeep Puligilla <spuligil@codeaurora.org>
Date: Wed, 20 Sep 2017 17:16:41 -0700
Subject: qcacld-3.0: Add vdev sanity check in action frame handler

Add vdev sanity check before accessing the wma
interface structure as part of the action
frame evt handling.

Change-Id: I19998b56f5932df92e2ab8b73bc17d9c60f33ed6
CRs-Fixed: 2113423
---
 core/wma/inc/wma_internal.h | 11 +++++++++++
 core/wma/src/wma_dev_if.c   |  6 +++++-
 core/wma/src/wma_main.c     |  5 +++++
 3 files changed, 21 insertions(+), 1 deletion(-)

diff --git a/core/wma/inc/wma_internal.h b/core/wma/inc/wma_internal.h
index 8b9c329..fa9a9af 100644
--- a/core/wma/inc/wma_internal.h
+++ b/core/wma/inc/wma_internal.h
@@ -1555,4 +1555,15 @@ QDF_STATUS wma_extract_single_phyerr_spectral(void *handle,
  */
 int wma_rx_aggr_failure_event_handler(void *handle, u_int8_t *event_buf,
 							u_int32_t len);
+
+/*
+ * wma_is_vdev_valid - check the vdev status
+ * @vdev_id: vdev identifier
+ *
+ * This function verifies the vdev validity
+ *
+ * Return: 'true' on valid vdev else 'false'
+ */
+bool wma_is_vdev_valid(uint32_t vdev_id);
+
 #endif
diff --git a/core/wma/src/wma_dev_if.c b/core/wma/src/wma_dev_if.c
index d0123b4..309c4e4 100644
--- a/core/wma/src/wma_dev_if.c
+++ b/core/wma/src/wma_dev_if.c
@@ -1120,13 +1120,17 @@ int wma_vdev_start_resp_handler(void *handle, uint8_t *cmd_param_info,
 	return 0;
 }
 
-static bool wma_is_vdev_valid(uint32_t vdev_id)
+bool wma_is_vdev_valid(uint32_t vdev_id)
 {
 	tp_wma_handle wma_handle = cds_get_context(QDF_MODULE_ID_WMA);
 
 	if (NULL == wma_handle)
 		return false;
 
+	/* No of interface are allocated based on max_bssid value */
+	if (vdev_id >= wma_handle->max_bssid)
+		return false;
+
 	return wma_handle->interfaces[vdev_id].vdev_active;
 }
 
diff --git a/core/wma/src/wma_main.c b/core/wma/src/wma_main.c
index a4be3ce..c1f767e 100644
--- a/core/wma/src/wma_main.c
+++ b/core/wma/src/wma_main.c
@@ -2128,6 +2128,11 @@ wma_action_frame_filter_mac_event_handler(void *handle, u_int8_t *event_buf,
 		return -EINVAL;
 	}
 
+	if (!wma_is_vdev_valid(event->vdev_id)) {
+		WMA_LOGE(FL("Invalid vdev id"));
+		return -EINVAL;
+	}
+
 	intr = &wma_handle->interfaces[event->vdev_id];
 	/* command is in progress */
 	if (!intr->action_frame_filter) {
-- 
cgit v1.1

